lib/jekyll/filters.rb

```diff
@@ -371,15 +371,26 @@ def time(input)
     def item_property(item, property)
       if item.respond_to?(:to_liquid)
         property.to_s.split(".").reduce(item.to_liquid) do |subvalue, attribute|
-          subvalue[attribute]
+          parse_sort_input(subvalue[attribute])
         end
       elsif item.respond_to?(:data)
-        item.data[property.to_s]
+        parse_sort_input(item.data[property.to_s])
       else
-        item[property.to_s]
+        parse_sort_input(item[property.to_s])
       end
     end
 
+    private
+    # return numeric values as numbers for proper sorting
+    def parse_sort_input(property)
+      if (Integer(property) && true rescue false)
+        property.to_i
+      elsif (Float(result) && true rescue false)
```

Comment:
No doubt, this appears elegant. But I'm on the fence regarding the use of inline `rescue` especially since its going to `rescue` everything to `false`

Is there another way to go about this, that doesn't involve coupling `sorting` with `rescue`-ing..?

Reply:
The first way I went about it was using regex to check if the string was a number, which would avoid the "rescue" but the Integer("string") and Float("string") with rescue seemed to be more popular and considered more "readable". I have no preference for either, although it seems from further reading that the regex approach is faster.

The regex approach would look as below. I believe it can be reduced to just one "if" since both integers and floats will evaluate as "true" and can both be converted to floats and still sort properly.

```ruby
def parse_sort_input(property)
  numeric = /\A-?(\d+)?\.?\d+\Z/
  if (property =~ numeric)
    return property.to_f
  end
  return property
end
```

Did you have another approach in mind?

Reply:
I like this approach.. the method's intention is straightforward and transparent.
Unless someone else comes up and tell us this method is slower than the previous, I'd go with this approach, with some minor modifications..:

like, why are we capturing just the digit before the decimal if we intend it to be a float?
```
%r!\A(-?\d+?\.?\d?)\Z!
```
would be better **IMO**.. it captures the whole numeric..
now the refactoring..

```ruby
def parse_sort_input(property)
  last_match = property.match(%r!\A(-?\d+?\.?\d?)\Z!)[0]
  return property if last_match.nil?
  last_match.to_f
end
```

And lastly, wrap this contribution with a test that confirms your code actually does what it advertises.. :smiley:

Reply:
Thanks! I am new to testing, so it might take me a little time to work through that.

In the meantime, a couple questions:
1. Should I wait to submit a new pull request when I have both the code updated and the test written?
2. I'm also new to regex (I mean, I'm new to all of this, haha) and I see that while I was returning the whole numeric by doing property.to_f, I was only checking the pattern up to the decimal.
3. In your improved regex, should the second d? still be d+? I run into a problem where 12.3 matches, but 12.34 does not.
3. With the refactored code, if 
```property.match(%r!\A(-?\d+?\.?\d?)\Z!) is nil (i.e. there is no match), then
```property.match(%r!\A(-?\d+?\.?\d?)\Z!)[0]
returns the error "undefined method `[]' for nil:NilClass"
Am I missing something, or should we check if property.match(regex) is nil before assigning property.match(regex)[0] to last_match?

This works for me:
```ruby
def parse_sort_inputb(property)
  last_match = property.match(%r!\A(-?\d+?\.?\d+?)\Z!)
  return property if last_match.nil?
  return last_match[0].to_f
end

Reply:
@greenkg Yes, you're right. There was a mistake in both my suggested regex and Ruby code.. and the fact that you were returning the entire float anyways..
Besides, matching a regex and then converting the match to a float was not a good suggestion, either.. :face_with_head_bandage: 

With the regex, my intention was to match a larger set (that includes `1`, `1.`, `1.0`, `1.01`, `1.001`..) but in my haste I used the wrong quantifier. the second `\d` should've been `\d*` or `\d*?`

Regarding the Ruby aspect, your code right above :point_up_2:  is fine..(the last `return` isn't necessary in Ruby). But the following is better..

```ruby
def parse_sort_input(property)
  return property unless property =~ %r!\A-?\d+?\.?\d*\Z!
  property.to_f
end
```

Reply:
`\d+?` is the same as `\d*`
Make sure the RegEx *does not* match the empty string, as the example above would.

Reply:
@pathawks When I test empty strings, using d+? does not match an empty string but using d* in both locations does. So @ashmaroli 's modified regex does not match "" when I test it for that, at least. (Although I wasn't thinking to check for that before you pointed it out!)

Reply:
I see: `\d+?` is the same as `\d+`, it seems the `?` is just ignored?

In this case, this RegEx does _not_ match the empty string (yay) but neither does it match `.5` (oops).

```irb
 2.4.0 :001 > "" =~ %r!\A-?\d+?\.?\d*\Z!
 => nil 
 2.4.0 :002 > ".5" =~ %r!\A-?\d+?\.?\d*\Z!
 => nil 
 2.4.0 :003 > "0.5" =~ %r!\A-?\d+?\.?\d*\Z!
 => 0 
```

Reply:
@pathawks Ah okay, I see. Shoot. Now I've got the following
```ruby
return property unless (property =~ %r!\A-?(\d*\.?\d*)(\.?\d+)\Z!)
```

Reply:
@ashmaroli When you have a moment, I'd love some input on adding the test. With the last code revision in my branch, I'm now failing a sort test but I *think* it *should* be failing this test now.

It is failing at test_filters.rb line 1039 for the following test:
```ruby
assert_equal(
          [{ "a" => "10" }, { "a" => "2" }],
          @filter.sort([{ "a" => "10" }, { "a" => "2" }], "a")
        )
```
I *think* this test is expecting the very result that this change is supposed to avoid. So "10" and "2" are now getting sorted as 10 and 2.

Am I understanding that right and should this test be removed (and then replaced with a new test showing the opposite, essentially?)

Thank you for all the help!

Reply:
I would suggest:

```ruby
NUMBER_LIKE = %r!\A\s*(?:\d+\.?\d*|\.?\d+)\s*\Z!
```

We don't care if the String has leading or trailing spaces. If there is a decimal point, there _must_ be either a numeral preceding or (if there is not a preceding numeral) there must be a following numeral.

This RegEx matches `"1"`, `" 1 "`, `".1"`, `"1."`, and `"1.1"`. It does _not_ match `"1 1"`, `""`, or `"."`.

By making this a named constant, we can make the code read a bit more easily:

```ruby
return Float(property) if property =~ NUMBER_LIKE
return property
```

Reply:
We can also clean the RegEx up once merged. I don't mean for us to get stuck on that :+1: